File: problem_010.c
// -*- compile-command: "gcc -std=c99 -o problem_010 problem_010.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 10:\n"
           "Find the sum of all the primes below two million.\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int limit = 2000000;
    int nprimes = 1.5 * limit / log(limit);    // Estimate number of primes under limit
    int *primes = malloc(nprimes*sizeof(int));
    unsigned long long sum = 0;

    primeSieve(primes, nprimes);

    for (int i=0; primes[i]<limit; i++)
        sum += primes[i];

    printf("Sum of primes: %llu\n", sum);
    printElapsedTime(start);

    free(primes);
    return 0;
}

File: problem_010.rs
// -*- compile-command: "rustc -o problem_010_rs problem_010.rs" -*-
// Copyright (c) 2016 Michael Caldwell
//use std::num;

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 {
        return true;
    }
    if n % 2 == 0 {
        return false;
    }

    let limit = (n as f64).sqrt() as u64;
    for i in 3..limit+1 {
        if n % i == 0 {
            return false;
        }
    }
    return true;
}


fn main() {
    println!("Project Euler - Problem 10");
    println!("Find the sum of all the primes below two million.\n");

    let mut sum: u64 = 0;

    for n in 2..1999999 {
        if is_prime(n) {
            sum += n
        }
    }

    println!("{}", sum);
}
